Add metadata for Jetty 12.1.3

Our [Jetty smoke test](https://github.com/spring-projects/spring-aot-smoke-tests/tree/main/boot/servlet-jetty) fails because of missing Jetty metadata.

We've worked around that by using `buildArgs.addAll('-H:+UnlockExperimentalVMOptions', '-H:Preserve=package=org.eclipse.jetty.*')`.

We're using `org.eclipse.jetty:jetty-server:12.1.3`.

```
org.springframework.context.ApplicationContextException: Unable to start web server
	at org.springframework.boot.web.server.servlet.context.ServletWebServerApplicationContext.onRefresh(ServletWebServerApplicationContext.java:167)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:614)
	at org.springframework.boot.web.server.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:143)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:765)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:454)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:321)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1374)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1363)
	at com.example.servlet.jetty.ServletJettyApplication.main(ServletJettyApplication.java:26)
	at java.base@25/java.lang.invoke.LambdaForm$DMH/sa346b79c.invokeStaticInit(LambdaForm$DMH)
Caused by: java.lang.IllegalStateException: java.lang.NoSuchMethodException: org.eclipse.jetty.ee11.servlet.security.ConstraintSecurityHandler.<init>()
	at org.eclipse.jetty.ee11.servlet.ServletContextHandler.newSecurityHandler(ServletContextHandler.java:1298)
	at org.eclipse.jetty.ee11.servlet.ServletContextHandler.getSecurityHandler(ServletContextHandler.java:1361)
	at org.eclipse.jetty.ee11.servlet.ServletContextHandler.relinkHandlers(ServletContextHandler.java:1009)
	at org.eclipse.jetty.ee11.servlet.ServletContextHandler.<init>(ServletContextHandler.java:296)
	at org.eclipse.jetty.ee11.webapp.WebAppContext.<init>(WebAppContext.java:193)
	at org.eclipse.jetty.ee11.webapp.WebAppContext.<init>(WebAppContext.java:147)
	at org.springframework.boot.jetty.servlet.JettyEmbeddedWebAppContext.<init>(JettyEmbeddedWebAppContext.java:33)
	at org.springframework.boot.jetty.servlet.JettyServletWebServerFactory.getWebServer(JettyServletWebServerFactory.java:158)
	at org.springframework.boot.web.server.servlet.context.ServletWebServerApplicationContext.createWebServer(ServletWebServerApplicationContext.java:190)
	at org.springframework.boot.web.server.servlet.context.ServletWebServerApplicationContext.onRefresh(ServletWebServerApplicationContext.java:164)
	... 9 more
Caused by: java.lang.NoSuchMethodException: org.eclipse.jetty.ee11.servlet.security.ConstraintSecurityHandler.<init>()
	at java.base@25/java.lang.Class.checkConstructor(DynamicHub.java:1356)
	at java.base@25/java.lang.Class.getConstructor0(DynamicHub.java:1565)
	at java.base@25/java.lang.Class.getDeclaredConstructor(DynamicHub.java:2491)
	at org.eclipse.jetty.ee11.servlet.ServletContextHandler.newSecurityHandler(ServletContextHandler.java:1294)
	... 18 more
```

--- a/metadata/org.eclipse.jetty/jetty-server/12.1.3/index.json
+++ b/metadata/org.eclipse.jetty/jetty-server/12.1.3/index.json
@@ -1,0 +1,4 @@
+[
+  "reflect-config.json",
+  "resource-config.json"
+]

--- a/metadata/org.eclipse.jetty/jetty-server/12.1.3/reflect-config.json
+++ b/metadata/org.eclipse.jetty/jetty-server/12.1.3/reflect-config.json
@@ -1,0 +1,838 @@
+[
+  {
+    "condition": {
+      "typeReachable": "org.eclipse.jetty.ee10.webapp.StandardDescriptorProcessor"
+    },
+    "name": "[Ljava.lang.String;"
+  },
+  {
+    "condition": {
+      "typeReachable": "org.eclipse.jetty.xml.BaseClassCatalog$CatalogReader"
+    },
+    "name": "com.sun.org.apache.xerces.internal.jaxp.DocumentBuilderFactoryImpl",
+    "methods": [
+      {
+        "name": "<init>",
+        "parameterTypes": []
+      }
+    ]
+  },
+  {
+    "condition": {
+      "typeReachable": "org.eclipse.jetty.xml.XmlParser"
+    },
+    "name": "com.sun.org.apache.xerces.internal.jaxp.SAXParserFactoryImpl",
+    "methods": [
+      {
+        "name": "<init>",
+        "parameterTypes": []
+      }
+    ]
+  },
+  {
+    "condition": {
+      "typeReachable": "org.eclipse.jetty.util.security.SecurityUtils"
+    },
+    "name": "java.lang.System",
+    "methods": [
+      {
+        "name": "getSecurityManager",
+        "parameterTypes": []
+      }
+    ]
+  },
+  {
+    "condition": {
+      "typeReachable": "org.eclipse.jetty.util.VirtualThreads"
+    },
+    "name": "java.lang.Thread",
+    "methods": [
+      {
+        "name": "isVirtual",
+        "parameterTypes": []
+      }
+    ]
+  },
+  {
+    "condition": {
+      "typeReachable": "org.eclipse.jetty.util.statistic.SampleStatistic"
+    },
+    "name": "java.lang.Thread",
+    "fields": [
+      {
+        "name": "threadLocalRandomProbe"
+      }
+    ]
+  },
+  {
+    "condition": {
+      "typeReachable": "org.eclipse.jetty.util.Uptime$DefaultImpl"
+    },
+    "name": "java.lang.management.ManagementFactory",
+    "methods": [
+      {
+        "name": "getRuntimeMXBean",
+        "parameterTypes": []
+      }
+    ]
+  },
+  {
+    "condition": {
+      "typeReachable": "org.eclipse.jetty.util.Uptime$DefaultImpl"
+    },
+    "name": "java.lang.management.RuntimeMXBean",
+    "methods": [
+      {
+        "name": "getUptime",
+        "parameterTypes": []
+      }
+    ]
+  },
+  {
+    "condition": {
+      "typeReachable": "org.eclipse.jetty.util.security.SecurityUtils"
+    },
+    "name": "java.security.AccessController",
+    "methods": [
+      {
+        "name": "doPrivileged",
+        "parameterTypes": [
+          "java.security.PrivilegedAction"
+        ]
+      }
+    ]
+  },
+  {
+    "condition": {
+      "typeReachable": "org.eclipse.jetty.session.DefaultSessionIdManager"
+    },
+    "name": "java.security.SecureRandomParameters"
+  },
+  {
+    "condition": {
+      "typeReachable": "org.eclipse.jetty.util.VirtualThreads"
+    },
+    "name": "java.util.concurrent.Executors",
+    "methods": [
+      {
+        "name": "newVirtualThreadPerTaskExecutor",
+        "parameterTypes": []
+      }
+    ]
+  },
+  {
+    "condition": {
+      "typeReachable": "org.eclipse.jetty.util.statistic.SampleStatistic"
+    },
+    "name": "java.util.concurrent.atomic.Striped64",
+    "fields": [
+      {
+        "name": "base"
+      },
+      {
+        "name": "cellsBusy"
+      }
+    ]
+  },
+  {
+    "condition": {
+      "typeReachable": "org.eclipse.jetty.util.security.SecurityUtils"
+    },
+    "name": "javax.security.auth.Subject",
+    "methods": [
+      {
+        "name": "doAs",
+        "parameterTypes": [
+          "javax.security.auth.Subject",
+          "java.security.PrivilegedAction"
+        ]
+      }
+    ]
+  },
+  {
+    "condition": {
+      "typeReachable": "org.eclipse.jetty.ee10.servlet.ServletHolder"
+    },
+    "name": "jetty.HelloWorldServlet",
+    "methods": [
+      {
+        "name": "<init>",
+        "parameterTypes": []
+      }
+    ]
+  },
+  {
+    "condition": {
+      "typeReachable": "org.eclipse.jetty.ee10.servlet.ServletHolder"
+    },
+    "name": "org.eclipse.jetty.ee10.servlet.DefaultServlet",
+    "methods": [
+      {
+        "name": "<init>",
+        "parameterTypes": []
+      }
+    ]
+  },
+  {
+    "condition": {
+      "typeReachable": "org.eclipse.jetty.ee11.servlet.ServletHolder"
+    },
+    "name": "org.eclipse.jetty.ee11.servlet.DefaultServlet",
+    "methods": [
+      {
+        "name": "<init>",
+        "parameterTypes": []
+      }
+    ]
+  },
+  {
+    "condition": {
+      "typeReachable": "org.eclipse.jetty.ee10.servlet.ServletHolder"
+    },
+    "name": "org.eclipse.jetty.ee10.servlet.NoJspServlet",
+    "methods": [
+      {
+        "name": "<init>",
+        "parameterTypes": []
+      }
+    ]
+  },
+  {
+    "condition": {
+      "typeReachable": "org.eclipse.jetty.ee11.servlet.ServletHolder"
+    },
+    "name": "org.eclipse.jetty.ee11.servlet.NoJspServlet",
+    "methods": [
+      {
+        "name": "<init>",
+        "parameterTypes": []
+      }
+    ]
+  },
+  {
+    "condition": {
+      "typeReachable": "org.eclipse.jetty.ee10.servlet.ListenerHolder"
+    },
+    "name": "org.eclipse.jetty.ee10.servlet.listener.IntrospectorCleaner",
+    "methods": [
+      {
+        "name": "<init>",
+        "parameterTypes": []
+      }
+    ]
+  },
+  {
+    "condition": {
+      "typeReachable": "org.eclipse.jetty.ee11.servlet.ListenerHolder"
+    },
+    "name": "org.eclipse.jetty.ee11.servlet.listener.IntrospectorCleaner",
+    "methods": [
+      {
+        "name": "<init>",
+        "parameterTypes": []
+      }
+    ]
+  },
+  {
+    "condition": {
+      "typeReachable": "org.eclipse.jetty.ee10.servlet.ServletContextHandler"
+    },
+    "name": "org.eclipse.jetty.ee10.servlet.security.ConstraintSecurityHandler",
+    "methods": [
+      {
+        "name": "<init>",
+        "parameterTypes": []
+      }
+    ]
+  },
+  {
+    "condition": {
+      "typeReachable": "org.eclipse.jetty.ee11.servlet.ServletContextHandler"
+    },
+    "name": "org.eclipse.jetty.ee11.servlet.security.ConstraintSecurityHandler",
+    "methods": [
+      {
+        "name": "<init>",
+        "parameterTypes": []
+      }
+    ]
+  },
+  {
+    "condition": {
+      "typeReachable": "org.eclipse.jetty.ee10.webapp.ServletsConfiguration"
+    },
+    "name": "org.eclipse.jetty.ee10.servlets.PushCacheFilter"
+  },
+  {
+    "condition": {
+      "typeReachable": "org.eclipse.jetty.util.IncludeExcludeSet"
+    },
+    "name": "org.eclipse.jetty.ee10.webapp.ClassMatcher$ByLocationOrModule",
+    "methods": [
+      {
+        "name": "<init>",
+        "parameterTypes": []
+      }
+    ]
+  },
+  {
+    "condition": {
+      "typeReachable": "org.eclipse.jetty.util.IncludeExcludeSet"
+    },
+    "name": "org.eclipse.jetty.ee10.webapp.ClassMatcher$ByPackageOrName",
+    "methods": [
+      {
+        "name": "<init>",
+        "parameterTypes": []
+      }
+    ]
+  },
+  {
+    "condition": {
+      "typeReachable": "org.eclipse.jetty.ee10.webapp.IterativeDescriptorProcessor"
+    },
+    "name": "org.eclipse.jetty.ee10.webapp.StandardDescriptorProcessor",
+    "methods": [
+      {
+        "name": "visitListener",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee10.webapp.WebAppContext",
+          "org.eclipse.jetty.ee10.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitLocaleEncodingList",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee10.webapp.WebAppContext",
+          "org.eclipse.jetty.ee10.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitSecurityConstraint",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee10.webapp.WebAppContext",
+          "org.eclipse.jetty.ee10.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitServlet",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee10.webapp.WebAppContext",
+          "org.eclipse.jetty.ee10.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitServletMapping",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee10.webapp.WebAppContext",
+          "org.eclipse.jetty.ee10.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitSessionConfig",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee10.webapp.WebAppContext",
+          "org.eclipse.jetty.ee10.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitWelcomeFileList",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee10.webapp.WebAppContext",
+          "org.eclipse.jetty.ee10.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      }
+    ]
+  },
+  {
+    "condition": {
+      "typeReachable": "org.eclipse.jetty.ee10.webapp.StandardDescriptorProcessor"
+    },
+    "name": "org.eclipse.jetty.ee10.webapp.StandardDescriptorProcessor",
+    "methods": [
+      {
+        "name": "visitContextParam",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee10.webapp.WebAppContext",
+          "org.eclipse.jetty.ee10.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitDefaultContextPath",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee10.webapp.WebAppContext",
+          "org.eclipse.jetty.ee10.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitDenyUncoveredHttpMethods",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee10.webapp.WebAppContext",
+          "org.eclipse.jetty.ee10.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitDisplayName",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee10.webapp.WebAppContext",
+          "org.eclipse.jetty.ee10.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitErrorPage",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee10.webapp.WebAppContext",
+          "org.eclipse.jetty.ee10.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitFilter",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee10.webapp.WebAppContext",
+          "org.eclipse.jetty.ee10.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitFilterMapping",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee10.webapp.WebAppContext",
+          "org.eclipse.jetty.ee10.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitJspConfig",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee10.webapp.WebAppContext",
+          "org.eclipse.jetty.ee10.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitListener",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee10.webapp.WebAppContext",
+          "org.eclipse.jetty.ee10.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitLocaleEncodingList",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee10.webapp.WebAppContext",
+          "org.eclipse.jetty.ee10.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitLoginConfig",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee10.webapp.WebAppContext",
+          "org.eclipse.jetty.ee10.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitMimeMapping",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee10.webapp.WebAppContext",
+          "org.eclipse.jetty.ee10.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitRequestCharacterEncoding",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee10.webapp.WebAppContext",
+          "org.eclipse.jetty.ee10.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitResponseCharacterEncoding",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee10.webapp.WebAppContext",
+          "org.eclipse.jetty.ee10.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitSecurityConstraint",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee10.webapp.WebAppContext",
+          "org.eclipse.jetty.ee10.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitSecurityRole",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee10.webapp.WebAppContext",
+          "org.eclipse.jetty.ee10.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitServlet",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee10.webapp.WebAppContext",
+          "org.eclipse.jetty.ee10.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitServletMapping",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee10.webapp.WebAppContext",
+          "org.eclipse.jetty.ee10.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitSessionConfig",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee10.webapp.WebAppContext",
+          "org.eclipse.jetty.ee10.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitTagLib",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee10.webapp.WebAppContext",
+          "org.eclipse.jetty.ee10.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitWelcomeFileList",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee10.webapp.WebAppContext",
+          "org.eclipse.jetty.ee10.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      }
+    ]
+  },
+  {
+    "condition": {
+      "typeReachable": "org.eclipse.jetty.ee11.webapp.StandardDescriptorProcessor"
+    },
+    "name": "org.eclipse.jetty.ee11.webapp.StandardDescriptorProcessor",
+    "methods": [
+      {
+        "name": "visitContextParam",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee11.webapp.WebAppContext",
+          "org.eclipse.jetty.ee11.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitDefaultContextPath",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee11.webapp.WebAppContext",
+          "org.eclipse.jetty.ee11.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitDenyUncoveredHttpMethods",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee11.webapp.WebAppContext",
+          "org.eclipse.jetty.ee11.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitDisplayName",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee11.webapp.WebAppContext",
+          "org.eclipse.jetty.ee11.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitErrorPage",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee11.webapp.WebAppContext",
+          "org.eclipse.jetty.ee11.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitFilter",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee11.webapp.WebAppContext",
+          "org.eclipse.jetty.ee11.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitFilterMapping",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee11.webapp.WebAppContext",
+          "org.eclipse.jetty.ee11.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitJspConfig",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee11.webapp.WebAppContext",
+          "org.eclipse.jetty.ee11.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitListener",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee11.webapp.WebAppContext",
+          "org.eclipse.jetty.ee11.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitLocaleEncodingList",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee11.webapp.WebAppContext",
+          "org.eclipse.jetty.ee11.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitLoginConfig",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee11.webapp.WebAppContext",
+          "org.eclipse.jetty.ee11.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitMimeMapping",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee11.webapp.WebAppContext",
+          "org.eclipse.jetty.ee11.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitRequestCharacterEncoding",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee11.webapp.WebAppContext",
+          "org.eclipse.jetty.ee11.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitResponseCharacterEncoding",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee11.webapp.WebAppContext",
+          "org.eclipse.jetty.ee11.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitSecurityConstraint",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee11.webapp.WebAppContext",
+          "org.eclipse.jetty.ee11.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitSecurityRole",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee11.webapp.WebAppContext",
+          "org.eclipse.jetty.ee11.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitServlet",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee11.webapp.WebAppContext",
+          "org.eclipse.jetty.ee11.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitServletMapping",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee11.webapp.WebAppContext",
+          "org.eclipse.jetty.ee11.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitSessionConfig",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee11.webapp.WebAppContext",
+          "org.eclipse.jetty.ee11.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitTagLib",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee11.webapp.WebAppContext",
+          "org.eclipse.jetty.ee11.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      },
+      {
+        "name": "visitWelcomeFileList",
+        "parameterTypes": [
+          "org.eclipse.jetty.ee11.webapp.WebAppContext",
+          "org.eclipse.jetty.ee11.webapp.Descriptor",
+          "org.eclipse.jetty.xml.XmlParser$Node"
+        ]
+      }
+    ]
+  },
+  {
+    "condition": {
+      "typeReachable": "org.eclipse.jetty.ee10.webapp.JaasConfiguration"
+    },
+    "name": "org.eclipse.jetty.security.jaas.JAASLoginService"
+  },
+  {
+    "condition": {
+      "typeReachable": "org.eclipse.jetty.util.TypeUtil"
+    },
+    "name": "org.eclipse.jetty.util.TypeUtil",
+    "methods": [
+      {
+        "name": "getClassLoaderLocation",
+        "parameterTypes": [
+          "java.lang.Class"
+        ]
+      },
+      {
+        "name": "getCodeSourceLocation",
+        "parameterTypes": [
+          "java.lang.Class"
+        ]
+      },
+      {
+        "name": "getModuleLocation",
+        "parameterTypes": [
+          "java.lang.Class"
+        ]
+      },
+      {
+        "name": "getSystemClassLoaderLocation",
+        "parameterTypes": [
+          "java.lang.Class"
+        ]
+      }
+    ]
+  },
+  {
+    "condition": {
+      "typeReachable": "org.eclipse.jetty.util.security.SecurityUtils"
+    },
+    "name": "org.eclipse.jetty.util.security.SecurityUtils",
+    "methods": [
+      {
+        "name": "callableToPrivilegedAction",
+        "parameterTypes": [
+          "java.util.concurrent.Callable"
+        ]
+      }
+    ]
+  },
+  {
+    "condition": {
+      "typeReachable": "org.eclipse.jetty.session.DefaultSessionIdManager"
+    },
+    "name": "sun.security.provider.NativePRNG",
+    "methods": [
+      {
+        "name": "<init>",
+        "parameterTypes": []
+      },
+      {
+        "name": "<init>",
+        "parameterTypes": [
+          "java.security.SecureRandomParameters"
+        ]
+      }
+    ]
+  },
+  {
+    "condition": {
+      "typeReachable": "org.eclipse.jetty.server.ForwardedRequestCustomizer"
+    },
+    "name": "org.eclipse.jetty.server.ForwardedRequestCustomizer$Forwarded",
+    "methods": [
+      {
+        "name": "handleRFC7239",
+        "parameterTypes": [
+          "org.eclipse.jetty.http.HttpField"
+        ]
+      },
+      {
+        "name": "handleForwardedHost",
+        "parameterTypes": [
+          "org.eclipse.jetty.http.HttpField"
+        ]
+      },
+      {
+        "name": "handleForwardedFor",
+        "parameterTypes": [
+          "org.eclipse.jetty.http.HttpField"
+        ]
+      },
+      {
+        "name": "handleForwardedPort",
+        "parameterTypes": [
+          "org.eclipse.jetty.http.HttpField"
+        ]
+      },
+      {
+        "name": "handleProto",
+        "parameterTypes": [
+          "org.eclipse.jetty.http.HttpField"
+        ]
+      },
+      {
+        "name": "handleHttps",
+        "parameterTypes": [
+          "org.eclipse.jetty.http.HttpField"
+        ]
+      },
+      {
+        "name": "handleForwardedServer",
+        "parameterTypes": [
+          "org.eclipse.jetty.http.HttpField"
+        ]
+      },
+      {
+        "name": "handleCipherSuite",
+        "parameterTypes": [
+          "org.eclipse.jetty.http.HttpField"
+        ]
+      },
+      {
+        "name": "handleSslSessionId",
+        "parameterTypes": [
+          "org.eclipse.jetty.http.HttpField"
+        ]
+      }
+    ]
+  }
+]

--- a/metadata/org.eclipse.jetty/jetty-server/12.1.3/resource-config.json
+++ b/metadata/org.eclipse.jetty/jetty-server/12.1.3/resource-config.json
@@ -1,0 +1,278 @@
+{
+  "resources": {
+    "includes": [
+      {
+        "condition": {
+          "typeReachable": "org.eclipse.jetty.xml.BaseClassCatalog$CatalogReader"
+        },
+        "pattern": "\\Qjakarta/servlet/resources/j2ee_1_4.xsd\\E"
+      },
+      {
+        "condition": {
+          "typeReachable": "org.eclipse.jetty.xml.BaseClassCatalog$CatalogReader"
+        },
+        "pattern": "\\Qjakarta/servlet/resources/j2ee_web_services_client_1_1.xsd\\E"
+      },
+      {
+        "condition": {
+          "typeReachable": "org.eclipse.jetty.xml.BaseClassCatalog$CatalogReader"
+        },
+        "pattern": "\\Qjakarta/servlet/resources/jakartaee_9.xsd\\E"
+      },
+      {
+        "condition": {
+          "typeReachable": "org.eclipse.jetty.xml.BaseClassCatalog$CatalogReader"
+        },
+        "pattern": "\\Qjakarta/servlet/resources/jakartaee_web_services_client_2_0.xsd\\E"
+      },
+      {
+        "condition": {
+          "typeReachable": "org.eclipse.jetty.xml.BaseClassCatalog$CatalogReader"
+        },
+        "pattern": "\\Qjakarta/servlet/resources/javaee_5.xsd\\E"
+      },
+      {
+        "condition": {
+          "typeReachable": "org.eclipse.jetty.xml.BaseClassCatalog$CatalogReader"
+        },
+        "pattern": "\\Qjakarta/servlet/resources/javaee_6.xsd\\E"
+      },
+      {
+        "condition": {
+          "typeReachable": "org.eclipse.jetty.xml.BaseClassCatalog$CatalogReader"
+        },
+        "pattern": "\\Qjakarta/servlet/resources/javaee_7.xsd\\E"
+      },
+      {
+        "condition": {
+          "typeReachable": "org.eclipse.jetty.xml.BaseClassCatalog$CatalogReader"
+        },
+        "pattern": "\\Qjakarta/servlet/resources/javaee_8.xsd\\E"
+      },
+      {
+        "condition": {
+          "typeReachable": "org.eclipse.jetty.xml.BaseClassCatalog$CatalogReader"
+        },
+        "pattern": "\\Qjakarta/servlet/resources/javaee_web_services_client_1_2.xsd\\E"
+      },
+      {
+        "condition": {
+          "typeReachable": "org.eclipse.jetty.xml.BaseClassCatalog$CatalogReader"
+        },
+        "pattern": "\\Qjakarta/servlet/resources/javaee_web_services_client_1_3.xsd\\E"
+      },
+      {
+        "condition": {
+          "typeReachable": "org.eclipse.jetty.xml.BaseClassCatalog$CatalogReader"
+        },
+        "pattern": "\\Qjakarta/servlet/resources/javaee_web_services_client_1_4.xsd\\E"
+      },
+      {
+        "condition": {
+          "typeReachable": "org.eclipse.jetty.xml.BaseClassCatalog$CatalogReader"
+        },
+        "pattern": "\\Qjakarta/servlet/resources/web-app_2_2.dtd\\E"
+      },
+      {
+        "condition": {
+          "typeReachable": "org.eclipse.jetty.xml.BaseClassCatalog$CatalogReader"
+        },
+        "pattern": "\\Qjakarta/servlet/resources/web-app_2_3.dtd\\E"
+      },
+      {
+        "condition": {
+          "typeReachable": "org.eclipse.jetty.xml.BaseClassCatalog$CatalogReader"
+        },
+        "pattern": "\\Qjakarta/servlet/resources/web-app_2_4.xsd\\E"
+      },
+      {
+        "condition": {
+          "typeReachable": "org.eclipse.jetty.xml.BaseClassCatalog$CatalogReader"
+        },
+        "pattern": "\\Qjakarta/servlet/resources/web-app_2_5.xsd\\E"
+      },
+      {
+        "condition": {
+          "typeReachable": "org.eclipse.jetty.xml.BaseClassCatalog$CatalogReader"
+        },
+        "pattern": "\\Qjakarta/servlet/resources/web-app_3_0.xsd\\E"
+      },
+      {
+        "condition": {
+          "typeReachable": "org.eclipse.jetty.xml.BaseClassCatalog$CatalogReader"
+        },
+        "pattern": "\\Qjakarta/servlet/resources/web-app_3_1.xsd\\E"
+      },
+      {
+        "condition": {
+          "typeReachable": "org.eclipse.jetty.xml.BaseClassCatalog$CatalogReader"
+        },
+        "pattern": "\\Qjakarta/servlet/resources/web-app_4_0.xsd\\E"
+      },
+      {
+        "condition": {
+          "typeReachable": "org.eclipse.jetty.xml.BaseClassCatalog$CatalogReader"
+        },
+        "pattern": "\\Qjakarta/servlet/resources/web-app_5_0.xsd\\E"
+      },
+      {
+        "condition": {
+          "typeReachable": "org.eclipse.jetty.xml.BaseClassCatalog$CatalogReader"
+        },
+        "pattern": "\\Qjakarta/servlet/resources/web-app_6_0.xsd\\E"
+      },
+      {
+        "condition": {
+          "typeReachable": "org.eclipse.jetty.xml.BaseClassCatalog$CatalogReader"
+        },
+        "pattern": "\\Qjakarta/servlet/resources/web-app_6_1.xsd\\E"
+      },
+      {
+        "condition": {
+          "typeReachable": "org.eclipse.jetty.xml.BaseClassCatalog$CatalogReader"
+        },
+        "pattern": "\\Qjakarta/servlet/resources/web-common_3_0.xsd\\E"
+      },
+      {
+        "condition": {
+          "typeReachable": "org.eclipse.jetty.xml.BaseClassCatalog$CatalogReader"
+        },
+        "pattern": "\\Qjakarta/servlet/resources/web-common_3_1.xsd\\E"
+      },
+      {
+        "condition": {
+          "typeReachable": "org.eclipse.jetty.xml.BaseClassCatalog$CatalogReader"
+        },
+        "pattern": "\\Qjakarta/servlet/resources/web-common_4_0.xsd\\E"
+      },
+      {
+        "condition": {
+          "typeReachable": "org.eclipse.jetty.xml.BaseClassCatalog$CatalogReader"
+        },
+        "pattern": "\\Qjakarta/servlet/resources/web-common_5_0.xsd\\E"
+      },
+      {
+        "condition": {
+          "typeReachable": "org.eclipse.jetty.xml.BaseClassCatalog$CatalogReader"
+        },
+        "pattern": "\\Qjakarta/servlet/resources/web-common_6_0.xsd\\E"
+      },
+      {
+        "condition": {
+          "typeReachable": "org.eclipse.jetty.xml.BaseClassCatalog$CatalogReader"
+        },
+        "pattern": "\\Qjakarta/servlet/resources/web-common_6_1.xsd\\E"
+      },
+      {
+        "condition": {
+          "typeReachable": "org.eclipse.jetty.xml.BaseClassCatalog$CatalogReader"
+        },
+        "pattern": "\\Qjakarta/servlet/resources/web-fragment_3_0.xsd\\E"
+      },
+      {
+        "condition": {
+          "typeReachable": "org.eclipse.jetty.xml.BaseClassCatalog$CatalogReader"
+        },
+        "pattern": "\\Qjakarta/servlet/resources/web-fragment_3_1.xsd\\E"
+      },
+      {
+        "condition": {
+          "typeReachable": "org.eclipse.jetty.xml.BaseClassCatalog$CatalogReader"
+        },
+        "pattern": "\\Qjakarta/servlet/resources/web-fragment_4_0.xsd\\E"
+      },
+      {
+        "condition": {
+          "typeReachable": "org.eclipse.jetty.xml.BaseClassCatalog$CatalogReader"
+        },
+        "pattern": "\\Qjakarta/servlet/resources/web-fragment_5_0.xsd\\E"
+      },
+      {
+        "condition": {
+          "typeReachable": "org.eclipse.jetty.xml.BaseClassCatalog$CatalogReader"
+        },
+        "pattern": "\\Qjakarta/servlet/resources/web-fragment_6_0.xsd\\E"
+      },
+      {
+        "condition": {
+          "typeReachable": "org.eclipse.jetty.xml.BaseClassCatalog$CatalogReader"
+        },
+        "pattern": "\\Qjakarta/servlet/resources/web-fragment_6_1.xsd\\E"
+      },
+      {
+        "condition": {
+          "typeReachable": "org.eclipse.jetty.ee10.webapp.WebDescriptor$WebDescriptorParser"
+        },
+        "pattern": "\\Qorg/eclipse/jetty/ee10/webapp/catalog-ee10.xml\\E"
+      },
+      {
+        "condition": {
+          "typeReachable": "org.eclipse.jetty.ee11.webapp.WebDescriptor$WebDescriptorParser"
+        },
+        "pattern": "\\Qorg/eclipse/jetty/ee11/webapp/catalog-ee11.xml\\E"
+      },
+      {
+        "condition": {
+          "typeReachable": "org.eclipse.jetty.util.resource.ResourceFactory"
+        },
+        "pattern": "\\Qorg/eclipse/jetty/ee10/webapp/webdefault-ee10.xml\\E"
+      },
+      {
+        "condition": {
+          "typeReachable": "org.eclipse.jetty.util.resource.ResourceFactory"
+        },
+        "pattern": "\\Qorg/eclipse/jetty/ee11/webapp/webdefault-ee11.xml\\E"
+      },
+      {
+        "condition": {
+          "typeReachable": "org.eclipse.jetty.http.MimeTypes$1"
+        },
+        "pattern": "\\Qorg/eclipse/jetty/http/encoding.properties\\E"
+      },
+      {
+        "condition": {
+          "typeReachable": "org.eclipse.jetty.http.MimeTypes$1"
+        },
+        "pattern": "\\Qorg/eclipse/jetty/http/mime.properties\\E"
+      },
+      {
+        "condition": {
+          "typeReachable": "org.eclipse.jetty.server.Server"
+        },
+        "pattern": "\\Qorg/eclipse/jetty/server/jetty-dir.css\\E"
+      },
+      {
+        "condition": {
+          "typeReachable": "org.eclipse.jetty.util.Jetty"
+        },
+        "pattern": "\\Qorg/eclipse/jetty/version/build.properties\\E"
+      },
+      {
+        "condition": {
+          "typeReachable": "org.eclipse.jetty.util.resource.ResourceFactoryInternals"
+        },
+        "pattern": "\\Qorg/eclipse/jetty/version/build.properties\\E"
+      },
+      {
+        "condition": {
+          "typeReachable": "org.eclipse.jetty.xml.XmlParser"
+        },
+        "pattern": "\\Qorg/eclipse/jetty/xml/catalog-org.w3.xml\\E"
+      }
+    ]
+  },
+  "bundles": [
+    {
+      "name": "jakarta.servlet.LocalStrings",
+      "locales": [
+        "und"
+      ]
+    },
+    {
+      "name": "jakarta.servlet.http.LocalStrings",
+      "locales": [
+        "und"
+      ]
+    }
+  ]
+}

--- a/metadata/org.eclipse.jetty/jetty-server/index.json
+++ b/metadata/org.eclipse.jetty/jetty-server/index.json
@@ -2,6 +2,14 @@
   {
     "latest" : true,
     "module" : "org.eclipse.jetty:jetty-server",
+    "metadata-version" : "12.1.3",
+    "tested-versions" : [
+      "12.1.3"
+    ]
+  },
+  {
+    "module" : "org.eclipse.jetty:jetty-server",
+    "default-for" : "12\\.0\\..*",
     "metadata-version" : "12.0.1",
     "tested-versions" : [
       "12.0.1",
@@ -19,15 +27,7 @@
       "12.0.13",
       "12.0.14",
       "12.0.15",
-      "12.0.16",
-      "12.1.0.alpha0",
-      "12.1.0.alpha1",
-      "12.1.0.alpha2",
-      "12.1.0.beta0",
-      "12.1.0.beta1",
-      "12.1.0.beta2",
-      "12.1.0.beta3",
-      "12.1.0"
+      "12.0.16"
     ]
   },
   {

--- a/tests/src/index.json
+++ b/tests/src/index.json
@@ -406,6 +406,12 @@
     "name" : "org.eclipse.jetty:jetty-server",
     "versions" : [ "12.0.1" ]
   } ]
+ } , {
+  "test-project-path" : "org.eclipse.jetty/jetty-server/12.1.3",
+  "libraries" : [ {
+    "name" : "org.eclipse.jetty:jetty-server",
+    "versions" : [ "12.1.3" ]
+  } ]
 }, {
   "test-project-path" : "org.eclipse.jetty/jetty-util/12.0.9",
   "libraries" : [ {

--- a/tests/src/org.eclipse.jetty/jetty-server/12.1.3/.gitignore
+++ b/tests/src/org.eclipse.jetty/jetty-server/12.1.3/.gitignore
@@ -1,0 +1,4 @@
+gradlew
+gradlew.bat
+build/
+gradle/

--- a/tests/src/org.eclipse.jetty/jetty-server/12.1.3/README.md
+++ b/tests/src/org.eclipse.jetty/jetty-server/12.1.3/README.md
@@ -1,0 +1,3 @@
+# Jetty 12
+
+Metadata has been collected with the GraalVM agent.

--- a/tests/src/org.eclipse.jetty/jetty-server/12.1.3/build.gradle
+++ b/tests/src/org.eclipse.jetty/jetty-server/12.1.3/build.gradle
@@ -1,0 +1,38 @@
+/*
+ * Copyright and related rights waived via CC0
+ *
+ * You should have received a copy of the CC0 legalcode along with this
+ * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+ */
+plugins {
+    id "org.graalvm.internal.tck"
+}
+
+String libraryVersion = tck.testedLibraryVersion.get()
+
+dependencies {
+    testImplementation "org.eclipse.jetty:jetty-server:$libraryVersion"
+    testImplementation "org.eclipse.jetty.ee10:jetty-ee10-servlets:$libraryVersion"
+    testImplementation "org.eclipse.jetty.ee10:jetty-ee10-webapp:$libraryVersion"
+	testImplementation "org.eclipse.jetty.ee11:jetty-ee11-servlets:$libraryVersion"
+	testImplementation "org.eclipse.jetty.ee11:jetty-ee11-webapp:$libraryVersion"
+
+    testImplementation 'org.assertj:assertj-core:3.24.2'
+    testImplementation 'org.slf4j:slf4j-simple:2.0.9'
+}
+
+graalvmNative {
+    agent {
+        defaultMode = "conditional"
+        modes {
+            conditional {
+                userCodeFilterPath = "user-code-filter.json"
+            }
+        }
+        metadataCopy {
+            mergeWithExisting = false
+            inputTaskNames.add("test")
+            outputDirectories.add("src/test/resources/META-INF/native-image/org.eclipse.jetty/jetty-server")
+        }
+    }
+}

--- a/tests/src/org.eclipse.jetty/jetty-server/12.1.3/gradle.properties
+++ b/tests/src/org.eclipse.jetty/jetty-server/12.1.3/gradle.properties
@@ -1,0 +1,2 @@
+library.version=12.1.3
+metadata.dir=org.eclipse.jetty.ee1/jetty-ee11-servlet/12.1.3/

--- a/tests/src/org.eclipse.jetty/jetty-server/12.1.3/settings.gradle
+++ b/tests/src/org.eclipse.jetty/jetty-server/12.1.3/settings.gradle
@@ -1,0 +1,13 @@
+pluginManagement {
+    def tckPath = Objects.requireNonNullElse(
+            System.getenv("GVM_TCK_TCKDIR"),
+            "../../../../tck-build-logic"
+    )
+    includeBuild(tckPath)
+}
+
+plugins {
+    id "org.graalvm.internal.tck-settings" version "1.0.0-SNAPSHOT"
+}
+
+rootProject.name = 'jetty-tests'

--- a/tests/src/org.eclipse.jetty/jetty-server/12.1.3/src/test/java/jetty/HelloWorldServlet.java
+++ b/tests/src/org.eclipse.jetty/jetty-server/12.1.3/src/test/java/jetty/HelloWorldServlet.java
@@ -1,0 +1,23 @@
+/*
+ * Copyright and related rights waived via CC0
+ *
+ * You should have received a copy of the CC0 legalcode along with this
+ * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+ */
+package jetty;
+
+import jakarta.servlet.http.HttpServlet;
+import jakarta.servlet.http.HttpServletRequest;
+import jakarta.servlet.http.HttpServletResponse;
+
+import java.io.IOException;
+
+public class HelloWorldServlet extends HttpServlet {
+    @Override
+    protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws IOException {
+        resp.setStatus(200);
+        resp.setHeader("Content-Type", "text/plain");
+        resp.getWriter().print("Hello world");
+        resp.getWriter().flush();
+    }
+}

--- a/tests/src/org.eclipse.jetty/jetty-server/12.1.3/src/test/java/jetty/JettyTests.java
+++ b/tests/src/org.eclipse.jetty/jetty-server/12.1.3/src/test/java/jetty/JettyTests.java
@@ -1,0 +1,208 @@
+/*
+ * Copyright and related rights waived via CC0
+ *
+ * You should have received a copy of the CC0 legalcode along with this
+ * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+ */
+package jetty;
+
+import java.io.IOException;
+import java.net.ServerSocket;
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Path;
+import java.time.Duration;
+
+
+import org.eclipse.jetty.server.ConnectionFactory;
+import org.eclipse.jetty.server.Connector;
+import org.eclipse.jetty.server.ForwardedRequestCustomizer;
+import org.eclipse.jetty.server.Handler;
+import org.eclipse.jetty.server.HttpConfiguration;
+import org.eclipse.jetty.server.Request;
+import org.eclipse.jetty.server.Response;
+import org.eclipse.jetty.server.Server;
+import org.eclipse.jetty.util.Callback;
+import org.eclipse.jetty.util.TypeUtil;
+import org.eclipse.jetty.util.resource.ResourceFactory;
+import org.eclipse.jetty.util.resource.URLResourceFactory;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.io.TempDir;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class JettyTests {
+
+    private static int port;
+
+    private static final boolean DEBUG = false;
+
+    @BeforeAll
+    static void beforeAll() throws IOException {
+        port = findAvailablePort();
+        System.out.println("Using port " + port + " for Jetty");
+        System.setProperty("org.slf4j.simpleLogger.defaultLogLevel", DEBUG ? "debug" : "warn");
+    }
+
+    private static int findAvailablePort() throws IOException {
+        try (ServerSocket socket = new ServerSocket()) {
+            socket.bind(null);
+            return socket.getLocalPort();
+        }
+    }
+
+    @Test
+    void typeUtilWorks() {
+        assertThat(TypeUtil.fromName("java.lang.String")).isEqualTo(String.class);
+    }
+
+    @Test
+    void http() throws Exception {
+        Server server = new Server(port);
+        server.setHandler(new Handler.Abstract() {
+            @Override
+            public boolean handle(Request request, Response response, Callback callback) {
+                response.setStatus(200);
+                response.getHeaders().add("Content-Type", "text/plain");
+                response.write(true, ByteBuffer.wrap("Hello world".getBytes(StandardCharsets.UTF_8)), callback);
+                return true;
+            }
+        });
+        server.start();
+        try {
+            HttpResponse<String> response = doHttpRequest();
+            assertThat(response.statusCode()).isEqualTo(200);
+            assertThat(response.body()).isEqualTo("Hello world");
+        } finally {
+            server.stop();
+        }
+    }
+
+    @Test
+    void servletEe10() throws Exception {
+        Server server = new Server(port);
+        org.eclipse.jetty.ee10.servlet.ServletContextHandler handler = new org.eclipse.jetty.ee10.servlet.ServletContextHandler();
+        handler.setContextPath("/");
+        handler.addServlet(HelloWorldServlet.class, "/*");
+        server.setHandler(handler);
+        server.start();
+        try {
+            HttpResponse<String> response = doHttpRequest();
+            assertThat(response.statusCode()).isEqualTo(200);
+            assertThat(response.body()).isEqualTo("Hello world");
+        } finally {
+            server.stop();
+        }
+    }
+
+    @Test
+    void servletEe11() throws Exception {
+        Server server = new Server(port);
+        org.eclipse.jetty.ee11.servlet.ServletContextHandler handler = new org.eclipse.jetty.ee11.servlet.ServletContextHandler();
+        handler.setContextPath("/");
+        handler.addServlet(HelloWorldServlet.class, "/*");
+        server.setHandler(handler);
+        server.start();
+        try {
+            HttpResponse<String> response = doHttpRequest();
+            assertThat(response.statusCode()).isEqualTo(200);
+            assertThat(response.body()).isEqualTo("Hello world");
+        } finally {
+            server.stop();
+        }
+    }
+
+    @Test
+    void webappEe10(@TempDir Path tempDir) throws Exception {
+        // This is needed to work in a native image, otherwise it fails with:
+        // java.lang.IllegalArgumentException: not an allowed scheme: resource:/org/eclipse/jetty/ee10/webapp/webdefault-ee10.xml
+        // See https://github.com/eclipse/jetty.project/issues/9116
+        ResourceFactory.registerResourceFactory("resource", new URLResourceFactory());
+
+        Server server = new Server(port);
+        org.eclipse.jetty.ee10.webapp.WebAppContext context = new org.eclipse.jetty.ee10.webapp.WebAppContext();
+        // EnvConfiguration and PlusConfiguration uses JNDI, which is not what we want to include
+        context.setBaseResourceAsPath(tempDir);
+        context.setContextPath("/");
+        context.getServletContext().addServlet("HelloWorld", HelloWorldServlet.class).addMapping("/*");
+        server.setHandler(context);
+        server.start();
+        try {
+            HttpResponse<String> response = doHttpRequest();
+            assertThat(response.statusCode()).isEqualTo(200);
+            assertThat(response.body()).isEqualTo("Hello world");
+        } finally {
+            server.stop();
+        }
+    }
+
+    @Test
+    void webappEe11(@TempDir Path tempDir) throws Exception {
+        // This is needed to work in a native image, otherwise it fails with:
+        // java.lang.IllegalArgumentException: not an allowed scheme: resource:/org/eclipse/jetty/ee10/webapp/webdefault-ee10.xml
+        // See https://github.com/eclipse/jetty.project/issues/9116
+        ResourceFactory.registerResourceFactory("resource", new URLResourceFactory());
+
+        Server server = new Server(port);
+        org.eclipse.jetty.ee11.webapp.WebAppContext context = new org.eclipse.jetty.ee11.webapp.WebAppContext();
+        // EnvConfiguration and PlusConfiguration uses JNDI, which is not what we want to include
+        context.setBaseResourceAsPath(tempDir);
+        context.setContextPath("/");
+        context.getServletContext().addServlet("HelloWorld", HelloWorldServlet.class).addMapping("/*");
+        server.setHandler(context);
+        server.start();
+        try {
+            HttpResponse<String> response = doHttpRequest();
+            assertThat(response.statusCode()).isEqualTo(200);
+            assertThat(response.body()).isEqualTo("Hello world");
+        } finally {
+            server.stop();
+        }
+    }
+
+    @Test
+    void forwardHeaders() throws Exception {
+        Server server = new Server(port);
+        for (Connector connector : server.getConnectors()) {
+            for (ConnectionFactory connectionFactory : connector.getConnectionFactories()) {
+                if (connectionFactory instanceof HttpConfiguration.ConnectionFactory) {
+                    ((HttpConfiguration.ConnectionFactory) connectionFactory).getHttpConfiguration()
+                            .addCustomizer(new ForwardedRequestCustomizer());
+                }
+            }
+        }
+        server.setHandler(new Handler.Abstract() {
+            @Override
+            public boolean handle(Request request, Response response, Callback callback) {
+                response.setStatus(200);
+                response.getHeaders().add("Content-Type", "text/plain");
+                String content = "I am " + Request.getServerName(request) + ":" + Request.getServerPort(request);
+                response.write(true, ByteBuffer.wrap(content.getBytes(StandardCharsets.UTF_8)), callback);
+                return true;
+            }
+        });
+        server.start();
+        try {
+            HttpResponse<String> response = doHttpRequest("X-Forwarded-Host", "some-host", "X-Forwarded-Port", "12345");
+            assertThat(response.statusCode()).isEqualTo(200);
+            assertThat(response.body()).isEqualTo("I am some-host:12345");
+        } finally {
+            server.stop();
+        }
+    }
+
+    private static HttpResponse<String> doHttpRequest(String... headers) throws IOException, InterruptedException {
+        HttpClient client = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(1)).build();
+        HttpRequest.Builder request = HttpRequest.newBuilder(URI.create(String.format("http://localhost:%d/", port)))
+                .GET().header("Accept", "text/plain").timeout(Duration.ofSeconds(1));
+        if (headers.length > 0) {
+            request.headers(headers);
+        }
+        return client.send(request.build(), HttpResponse.BodyHandlers.ofString());
+    }
+}

--- a/tests/src/org.eclipse.jetty/jetty-server/12.1.3/src/test/resources/META-INF/native-image/test/reflect-config.json
+++ b/tests/src/org.eclipse.jetty/jetty-server/12.1.3/src/test/resources/META-INF/native-image/test/reflect-config.json
@@ -1,0 +1,6 @@
+[
+  {
+    "name": "jetty.HelloWorldServlet",
+    "allDeclaredConstructors": true
+  }
+]

--- a/tests/src/org.eclipse.jetty/jetty-server/12.1.3/user-code-filter.json
+++ b/tests/src/org.eclipse.jetty/jetty-server/12.1.3/user-code-filter.json
@@ -1,0 +1,10 @@
+{
+  "rules": [
+    {
+      "excludeClasses": "**"
+    },
+    {
+      "includeClasses": "org.eclipse.jetty.**"
+    }
+  ]
+}

